File: examples/example_voice_clone.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
FlowTTS Voice Clone Example
"""

import os
import json
import base64
from dotenv import load_dotenv
from tencentcloud.common import credential
from tencentcloud.common.profile.client_profile import ClientProfile
from tencentcloud.common.profile.http_profile import HttpProfile
from tencentcloud.trtc.v20190722 import trtc_client, models
from tencentcloud.common.exception.tencent_cloud_sdk_exception import TencentCloudSDKException

# Load environment variables from .env file
load_dotenv()

# ========== Configuration ==========
SDK_APP_ID = int(os.getenv("TENCENTCLOUD_SDK_APP_ID", "1400000000"))
MODEL = "flow_01_turbo"

# Voice clone settings
CLONE_AUDIO_FILE = "./test_data/clone_sample.wav"  # 16kHz mono WAV, 10-180 seconds
VOICE_NAME = "MyClonedVoice"
# ===================================


def create_client():
    """
    Create Tencent Cloud TRTC client with Keep-Alive enabled.
    Keep-Alive reuses TCP connections for better performance.
    """
    cred = credential.Credential(
        os.getenv("TENCENTCLOUD_SECRET_ID"),
        os.getenv("TENCENTCLOUD_SECRET_KEY")
    )

    http_profile = HttpProfile()
    http_profile.endpoint = os.getenv(
        "TENCENTCLOUD_ENDPOINT",
        "trtc.tencentcloudapi.com"
    )
    http_profile.reqTimeout = 120
    http_profile.keepAlive = True        # Enable Keep-Alive
    http_profile.pre_conn_pool_size = 3  # Connection pool size

    client_profile = ClientProfile()
    client_profile.httpProfile = http_profile

    return trtc_client.TrtcClient(cred, "ap-beijing", client_profile)


def voice_clone(client, audio_file, voice_name):
    """
    Clone voice from audio sample.

    Args:
        client: TRTC client instance
        audio_file: Path to audio file (16kHz mono WAV, 10-180 seconds)
        voice_name: Name for the cloned voice

    Returns:
        voice_id: Cloned voice ID
    """
    # Read and encode audio file
    with open(audio_file, 'rb') as f:
        audio_data = f.read()
    audio_base64 = base64.b64encode(audio_data).decode('utf-8')

    # Create clone request
    req = models.VoiceCloneRequest()
    params = {
        "Model": MODEL,
        "SdkAppId": SDK_APP_ID,
        "VoiceName": voice_name,
        "PromptAudio": audio_base64
    }
    req.from_json_string(json.dumps(params))

    print(f"Cloning voice: {voice_name}")

    # Execute cloning
    try:
        resp = client.VoiceClone(req)
        voice_id = resp.VoiceId

        print(f"Voice cloned successfully!")
        print(f"Voice ID: {voice_id}")
        print(f"\nNext: Use this voice_id in example_simple.py")
        print(f"Update VOICE_CONFIG:")
        print(f'  "VoiceId": "{voice_id}"')
        return voice_id

    except TencentCloudSDKException as e:
        print(f"Voice cloning failed!")
        print(f"Error Code: {e.code}")
        print(f"Error Message: {e.message}")
        print(f"Request ID: {e.requestId}")
        return None
    except Exception as e:
        print(f"Voice cloning failed: {e}")
        return None


def main():
    # Create client (reuse for multiple requests)
    client = create_client()

    # Clone voice from audio sample
    voice_clone(client, CLONE_AUDIO_FILE, VOICE_NAME)


if __name__ == "__main__":
    main()

File: examples/nodejs/example_voice_clone.js
#!/usr/bin/env node
/**
 * FlowTTS Voice Clone Example
 *
 * Corresponds to examples/python/example_voice_clone.py
 *
 * Note: Voice Clone and Non-streaming APIs use endpoint "trtc.tencentcloudapi.com",
 * while Streaming SSE API uses "trtc.ai.tencentcloudapi.com".
 */

import { readFileSync } from "fs";
import { resolve } from "path";
import { loadConfig, createClient, projectRoot } from "./utils.js";

// ========== Configuration ==========
const MODEL = "flow_01_turbo";
const CLONE_AUDIO_FILE = resolve(
  projectRoot(),
  "test_data/clone_sample.wav"
); // 16kHz mono WAV, 10-180 seconds
const VOICE_NAME = "MyClonedVoice";
// (Optional) Transcript of the reference audio, improves clone quality
const PROMPT_TEXT = "";
// (Optional) Language of the reference audio (ISO 639-1): zh/en/yue/ja/ko, default auto
const LANGUAGE = "";
// ====================================

async function voiceClone(client, audioFile, voiceName, sdkAppId) {
  // Read and encode audio file
  const audioData = readFileSync(audioFile);
  const audioBase64 = audioData.toString("base64");

  const params = {
    Model: MODEL,
    SdkAppId: sdkAppId,
    VoiceName: voiceName,
    PromptAudio: audioBase64,
  };
  // Optional: provide transcript of the reference audio for better quality
  if (PROMPT_TEXT) {
    params.PromptText = PROMPT_TEXT;
  }
  // Optional: specify language (ISO 639-1)
  if (LANGUAGE) {
    params.Language = LANGUAGE;
  }

  console.log(`Cloning voice: ${voiceName}`);

  try {
    const resp = await client.VoiceClone(params);
    const voiceId = resp.VoiceId;

    console.log(`Voice cloned successfully!`);
    console.log(`Voice ID: ${voiceId}`);
    console.log(`\nNext: Use this voice_id in example_streaming.js`);
    console.log(`Update VOICE_CONFIG:`);
    console.log(`  "VoiceId": "${voiceId}"`);
    return voiceId;
  } catch (e) {
    console.log(`Voice cloning failed!`);
    if (e.code) {
      console.log(`Error Code: ${e.code}`);
      console.log(`Error Message: ${e.message}`);
      console.log(`Request ID: ${e.requestId}`);
    } else {
      console.log(`Error: ${e.message}`);
    }
    return null;
  }
}

async function main() {
  const cfg = loadConfig();
  // Voice Clone API uses trtc.tencentcloudapi.com
  // (Streaming SSE API uses trtc.ai.tencentcloudapi.com)
  const client = createClient(cfg, "trtc.tencentcloudapi.com");

  await voiceClone(client, CLONE_AUDIO_FILE, VOICE_NAME, cfg.sdkAppId);
}

main().catch(console.error);
